pageserver/src/walredo.rs

```diff
@@ -538,165 +521,142 @@ impl PostgresRedoProcess {
-            // Send base image, if any. (If the record initializes the page, previous page
-            // version is not needed.)
-            timeout(
-                TIMEOUT,
-                stdin.write_all(&build_begin_redo_for_block_msg(tag)),
-            )
-            .await??;
-            if let Some(img) = base_img {
-                timeout(TIMEOUT, stdin.write_all(&build_push_page_msg(tag, &img))).await??;
+        while nresult < pg_constants::BLCKSZ.into() {
+            // If we have more data to write, wake up if 'stdin' becomes writeable or
+            // we have data to read. Otherwise only wake up if there's data to read.
+            let nfds = if nwrite < writebuf.len() { 3 } else { 2 };
+            nix::poll::poll(&mut pollfds[0..nfds], TIMEOUT.as_millis() as i32)?;
+
+            // If we have more data to write and 'stdin' is writeable, do write.
+            if nwrite < writebuf.len() && !pollfds[2].revents().unwrap().is_empty() {
+                nwrite += self.stdin.write(&writebuf[nwrite..])?;
             }
 
-            // Send WAL records.
-            for (lsn, rec) in records.iter() {
-                WAL_REDO_RECORD_COUNTER.inc();
-
-                stdin
-                    .write_all(&build_apply_record_msg(*lsn, &rec.rec))
-                    .await?;
-
-                //debug!("sent WAL record to wal redo postgres process ({:X}/{:X}",
-                //       r.lsn >> 32, r.lsn & 0xffff_ffff);
+            // If we have some data in stdout, read it to the result buffer.
+            if !pollfds[0].revents().unwrap().is_empty() {
+                nresult += self.stdout.read(&mut resultbuf[nresult..])?;
             }
-            //debug!("sent {} WAL records to wal redo postgres process ({:X}/{:X}",
-            //       records.len(), lsn >> 32, lsn & 0xffff_ffff);
-
-            // Send GetPage command to get the result back
-            timeout(TIMEOUT, stdin.write_all(&build_get_page_msg(tag))).await??;
-            timeout(TIMEOUT, stdin.flush()).await??;
-            //debug!("sent GetPage for {}", tag.blknum);
-            Ok::<(), Error>(())
-        };
-
-        // Read back new page image
-        let f_stdout = async {
-            let mut buf = [0u8; 8192];
 
-            timeout(TIMEOUT, stdout.read_exact(&mut buf)).await??;
-            //debug!("got response for {}", tag.blknum);
-            Ok::<[u8; 8192], Error>(buf)
-        };
-
-        let res = tokio::try_join!(f_stdout, f_stdin)?;
-
-        let buf = res.0;
+            // If we have some messages in stderr, forward them to the log.
+            if !pollfds[1].revents().unwrap().is_empty() {
+                let mut errbuf: [u8; 16384] = [0; 16384];
+                let n = self.stderr.read(&mut errbuf)?;
+
+                // The message might not be split correctly into lines here. But this is
+                // good enough, the important thing is to get the message to the log.
+                error!(
```

Comment:
Shouldn't we break here? Now wal-redo process failure will cause getting stuck in an infinite loop

Reply:
The WAL redo process should exit on error. But if it doesn't, there's also the timeout that will make it unstuck. If we had a 'break' here, then an innocent "elog(LOG)" would also cause the replay to fail.